### PR TITLE
fix: added input validation in admin registration

### LINKED ISSUE #1222: bug: no input validation in talawa admin registration
**Describe the bug**
While registering in talawa-admin, there is currently no input validation. Number, special characters, etc can also be used to register.
For example in `first name` and `last name` fields users can even input numbers and other characters instead of strings.

**To Reproduce**

1. Go to talawa-admin
2. Click on register button
3. Use any characters while registering in first name, last name fields
4. Check database new user will be registered.

**Expected behavior**
Numbers, specials characters and brackets should not be accepted as valid input from the user. Only strings should be allowed as valid input

**Actual behavior**
Currently there is no input validation in registration form

**Screenshots**

**Additional details**

https://github.com/PalisadoesFoundation/talawa-admin/assets/116940083/5749fa26-d54d-4ce4-a0e4-4d9ca4defb83

It can be observed that the `first name` and `last name` of a user can't be numbers, but still they are registered and added in database.

**Potential internship candidates**
Please read this if you are planning to apply for a Palisadoes Foundation internship https://github.com/PalisadoesFoundation/talawa/issues/359

### PR DESCRIPTION
**Pull request by MahendraDani**
**What kind of change does this PR introduce?**
bug fix

**Issue Number:**
fixes #1222 

**Did you add tests for your changes?**
No

**Snapshots/Videos:**

### When user registers with incorrect details

https://github.com/PalisadoesFoundation/talawa-admin/assets/116940083/6ac0a9a7-1242-4c4d-9770-f1aad8f5b0d5

- The user is not added in database and an toast with warning is given to the user.

### When user registers with valid details

https://github.com/PalisadoesFoundation/talawa-admin/assets/116940083/c1ed2d95-00d4-4578-b43a-38d329711260

- The user is successfully added in database if details are valid

**If relevant, did you update the documentation?**
No

**Summary**

1. Installed [Zod](https://zod.dev/) library for input validation.
2. Added validations on `firstname`, `lastname`, `email` and `password` fields.
3. If validation is success the user is successfully registered.
4. If validation fail the user given a warning to fill the details correctly.

**Does this PR introduce a breaking change?**
No

**Other information**

**Have you read the [contributing guide](https://github.com/PalisadoesFoundation/talawa-admin/blob/master/CONTRIBUTING.md)?**

Yes

### CODE REVIEW

**Comment by Kevoniat** [comment]
@MahendraDani  The Code Coverage  is down  please Investigate

**Comment by MahendraDani** [comment]
@Kevoniat Added test coverage for the changes I made. Please let me know if there are any changes.
Thanks
image

File: package.json [hunk 1]
```diff
@@ -66,7 +66,8 @@
     "typedoc-plugin-markdown": "^3.16.0",
     "typescript": "^4.3.5",
     "web-vitals": "^1.0.1",
-    "yarn": "^1.22.17"
+    "yarn": "^1.22.17",
+    "zod": "^3.22.4"
```
Comment (noman2002) [inline package.json:70]: Why this package is added ?

Comment (MahendraDani) [reply package.json:70]: @noman2002 [Zod](https://zod.dev/) is used for validation of input given by the user.

**Comment by MahendraDani** [comment]
@palisadoes Would you please review this PR?

**Comment by palisadoes** [comment]
There are reviewed already signed. Please work with them

**Comment by aashimawadhwa** [comment]
@MahendraDani  please add proper validation messages the message provided right now is very generic. the validation should work in hierarchy that is the warning should be the first name cannot be number and so on .

File: package.json (hunk 1 above)
Comment (noman2002) [reply package.json:70]: But we already have validation in various places without any package and its a very basic thing. I think it is not required to use any package for validation.

@aashimawadhwa your thoughts on this.

Comment (aashimawadhwa) [reply package.json:70]: yes i think we can make validation work without the zod package too. @MahendraDani @noman2002

Comment (MahendraDani) [reply package.json:70]: > yes i think we can make validation work without the zod package too. @MahendraDani @noman2002

@aashimawadhwa @noman2002 
I guess you are right. Should I close this PR?

Comment (MahendraDani) [reply package.json:70]: @aashimawadhwa @noman2002 
My aim was to validate input for special characters and other characters like `[`, `]`, `{`, `}`,`@`, `#` etc. 
But it seems that when we validate any input by string `zod` considers the above characters in string too so using `zod` does not help that much.

Comment (rishav-jha-mech) [reply package.json:70]: @MahendraDani why are you thinking on closing this PR ?, you can just remove zod and use TS or Regex to validate the strings

Comment (MahendraDani) [reply package.json:70]: Will it be fine if I use regex and remove the package `zod` ?

Comment (rishav-jha-mech) [reply package.json:70]: Yeah sure go ahead !

**Comment by MahendraDani** [comment]
> @MahendraDani please add proper validation messages the message provided right now is very generic. the validation should work in hierarchy that is the warning should be the first name cannot be number and so on .

I will remove `zod` package and use regex for validation and add more clear messages.
cc @aashimawadhwa @rishav-jha-mech

**Comment by MahendraDani** [comment]
Currently we do not have any validation for `password`, should I also add the following validation for `password`:
1. Length should be more than 9
2. Should contain atleast one lowercase character: `a-z`
3. Should contain atleast one uppercase character : `A-Z`
4. Should contain atleast one special character : `!@#$^&*`
5. Should contain atleast one number :`0-9`

cc @aashimawadhwa @rishav-jha-mech

**Comment by rishav-jha-mech** [comment]
Password length >= 6 is okay, if it's greater than 9, people will tend to forget their password, other validation steps are good to go

**Comment by MahendraDani** [comment]
> Password length >= 6 is okay, if it's greater than 9, people will tend to forget their password, other validation steps are good to go

Alright, Thanks

**Comment by MahendraDani** [comment]
### Changes introduced in https://github.com/PalisadoesFoundation/talawa-admin/pull/1247/commits/8b36c82c3ffc04a3c0d305eda1af6634c7b9d37d
- Removed `zod` package and used `Regular Expressions` for input validation
- Added input validation for `firstname`, `lastname`, `email`,`password`.
- Added more specific warning message for validation failures.

**Comment by MahendraDani** [comment]
After resolving merge conflicts when I am trying to commit changes I am getting this following error:
```
/bin/sh: markdown-toc: command not found
node:internal/errors:932
  const err = new Error(message);
              ^

Error: Command failed: markdown-toc -i "CODE_OF_CONDUCT.md" --bullets "-"
    at checkExecSyncError (node:child_process:890:11)
    at execSync (node:child_process:962:15)
    at file:///Users/mahendradani/Documents/Programming/GSOC24/Palisadoes/talawa-admin/scripts/githooks/update-toc.js:10:3
    at Array.forEach (<anonymous>)
    at file:///Users/mahendradani/Documents/Programming/GSOC24/Palisadoes/talawa-admin/scripts/githooks/update-toc.js:8:15
    at ModuleJob.run (node:internal/modules/esm/module_job:218:25)
    at async ModuleLoader.import (node:internal/modules/esm/loader:329:24)
    at async loadESM (node:internal/process/esm_loader:34:7)
    at async handleMainPromise (node:internal/modules/run_main:113:12) {
  status: 127,
  signal: null,
  output: [ null, null, null ],
  pid: 18954,
  stdout: null,
  stderr: null
}

Node.js v20.10.0
husky - pre-commit hook exited with code 1 (error)
```

cc @aashimawadhwa @rishav-jha-mech

**Comment by MahendraDani** [comment]
> After resolving merge conflicts when I am trying to commit changes I am getting this following error:
> 
> ```
> /bin/sh: markdown-toc: command not found
> node:internal/errors:932
>   const err = new Error(message);
>               ^
> 
> Error: Command failed: markdown-toc -i "CODE_OF_CONDUCT.md" --bullets "-"
>     at checkExecSyncError (node:child_process:890:11)
>     at execSync (node:child_process:962:15)
>     at file:///Users/mahendradani/Documents/Programming/GSOC24/Palisadoes/talawa-admin/scripts/githooks/update-toc.js:10:3
>     at Array.forEach (<anonymous>)
>     at file:///Users/mahendradani/Documents/Programming/GSOC24/Palisadoes/talawa-admin/scripts/githooks/update-toc.js:8:15
>     at ModuleJob.run (node:internal/modules/esm/module_job:218:25)
>     at async ModuleLoader.import (node:internal/modules/esm/loader:329:24)
>     at async loadESM (node:internal/process/esm_loader:34:7)
>     at async handleMainPromise (node:internal/modules/run_main:113:12) {
>   status: 127,
>   signal: null,
>   output: [ null, null, null ],
>   pid: 18954,
>   stdout: null,
>   stderr: null
> }
> 
> Node.js v20.10.0
> husky - pre-commit hook exited with code 1 (error)
> ```
> 
> cc @aashimawadhwa @rishav-jha-mech

I think the above error is because of the following command in `pacakage.json` file
```
"update:toc": "node scripts/githooks/update-toc.js"
```

File: package-lock.json [hunk 2]
```diff
@@ -25150,9 +25150,9 @@
       }
     },
     "node_modules/zod": {
-      "version": "3.21.4",
-      "resolved": "https://registry.npmjs.org/zod/-/zod-3.21.4.tgz",
-      "integrity": "sha512-m46AKbrzKVzOzs/DZgVnG5H55N1sv1M8qZU3A8RIKbs3mrACDNeIOeilDymVb2HdmP8uwshOCF4uJ8uM9rCqJw==",
+      "version": "3.22.4",
```
Comment (rishav-jha-mech) [inline package-lock.json:25153]: Revert change in this file

File: public/locales/en.json [hunk 3]
```diff
@@ -9,7 +9,11 @@
     "lastName": "Last Name",
     "email": "Email",
     "password": "Password",
-    "atleast_8_char_long": "Atleast 8 Character long",
+    "atleast_6_char_long": "Atleast 6 Character long",
+    "firstName_invalid": "First name should contain only lower and upper case letters",
+    "lastName_invalid": "Last name should contain only lower and upper case letters",
```
Comment (rishav-jha-mech) [inline public/locales/en.json:14]: Add the same messages in other languages too

File: package-lock.json (hunk 2 above)
Comment (MahendraDani) [reply package-lock.json:25153]: @rishav-jha-mech It is not working after reverting the files
I also tried to create a new branch and commit changes in that branch but it is also giving same error message

Comment (rishav-jha-mech) [reply package-lock.json:25153]: @MahendraDani in any case if we are not using zod, there is no point in upgrading its version in `package-lock.json`, you are saying its not working after changing the files, what do you mean by that ?, also have you tried deleting the `node_modules` folder and reinstalling packages ?

**Comment by rishav-jha-mech** [comment]
@MahendraDani please remove the merge conflicts

File: package-lock.json (hunk 2 above)
Comment (MahendraDani) [reply package-lock.json:25153]: @rishav-jha-mech I first unistalled `zod` using `npm uninstall --legacy-peer-deps zod` and then made the changes in `LoginPage.tsx` file.

**Pull request closed without merging**